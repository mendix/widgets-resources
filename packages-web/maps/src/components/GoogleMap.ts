import { Component, createElement } from "react";
import * as classNames from "classnames";
import ReactResizeDetector from "react-resize-detector";

import { Alert } from "./Alert";
import googleApiWrapper from "./GoogleApi";
import { Container, MapUtils } from "../utils/namespace";
import Utils from "../utils/Utils";
import { validLocation } from "../utils/Validations";

type MapProps = Container.MapProps;
type DataSourceLocationProps = Container.DataSourceLocationProps;
type Location = Container.Location;
type SharedProps = MapUtils.SharedProps;

export interface GoogleMapsProps extends SharedProps, MapProps {
    scriptsLoaded?: boolean;
    onClickMarker?: (event: google.maps.MouseEvent, locationAttr: DataSourceLocationProps) => void;
}

export interface GoogleMapState {
    center: google.maps.LatLngLiteral;
    alertMessage?: string;
    resized: boolean;
}

export class GoogleMap extends Component<GoogleMapsProps, GoogleMapState> {
    private map!: google.maps.Map;
    private defaultCenterLocation: google.maps.LatLngLiteral = { lat: 51.9066346, lng: 4.4861703 };
    private markers: google.maps.Marker[] = [];
    private bounds!: google.maps.LatLngBounds;
    private googleMapsNode?: HTMLDivElement;
    private readonly onResizeHandle = this.onResize.bind(this);

    readonly state: GoogleMapState = {
        center: this.getDefaultCenter(this.props),
        alertMessage: this.props.alertMessage,
        resized: false
    };

    render() {
        return createElement("div", {
                className: classNames("widget-maps", this.props.className),
                style: { ...this.props.divStyles, ...Utils.getDimensions(this.props) }
            },
            createElement(Alert, {
                bootstrapStyle: "danger",
                className: "widget-google-maps-alert"
            }, this.state.alertMessage),
            createElement("div",
                {
                    className: "widget-google-maps-wrapper"
                },
                createElement("div", {
                    className: "widget-google-maps",
                    ref: this.getRef
                })
            ),
            createElement(ReactResizeDetector, {
                handleWidth: true,
                handleHeight: true,
                refreshRate: 100,
                refreshMode: "throttle",
                onResize: this.onResizeHandle
            })
        );
    }

    componentDidMount() {
        if (this.props.scriptsLoaded) {
            this.createUpdateMap(this.props);
        }
    }

    componentWillReceiveProps(nextProps: GoogleMapsProps) {
        if (nextProps.alertMessage !== this.props.alertMessage) {
            this.setState({ alertMessage: nextProps.alertMessage });
        }
        if (nextProps.scriptsLoaded && (this.props.allLocations !== nextProps.allLocations || !this.map)) {
            this.createUpdateMap(nextProps);
            this.setState({ resized: false });
        }
    }

    componentDidUpdate() {
        if (this.map && !this.props.fetchingData && !this.state.resized) {
            this.map.setCenter(this.state.center);
        }
    }

    private getDefaultCenter(props: GoogleMapsProps): google.maps.LatLngLiteral {
        const { defaultCenterLatitude, defaultCenterLongitude } = props;
        const location = {
            latitude: defaultCenterLatitude && parseFloat(defaultCenterLatitude),
            longitude: defaultCenterLongitude && parseFloat(defaultCenterLongitude)
        };
        if (validLocation(location)) {
            return {
                lat: parseFloat(defaultCenterLatitude!),
                lng: parseFloat(defaultCenterLongitude!)
            };
        }

        return this.defaultCenterLocation;
    }

    private onResize() {
        // When map is placed on the non default tab, the maps has no size.
        // Therefor it need to update on the first view.
        if (this.map && !this.state.resized) {
            this.createUpdateMap(this.props);
            this.setState({ resized: true });
        }
    }

    private getRef = (node: HTMLDivElement) => {
        this.googleMapsNode = node;
    }

    private createUpdateMap = (props: GoogleMapsProps) => {
        const mapOptions = {
            zoom: props.zoomLevel,
            zoomControl: props.optionZoomControl,
            scrollwheel: props.optionScroll,
            draggable: props.optionDrag,
            streetViewControl: props.optionStreetView,
            mapTypeControl: props.mapTypeControl,
            fullscreenControl: props.fullScreenControl,
            rotateControl: props.rotateControl,
            minZoom: 2,
            maxZoom: 20,
            styles: this.getMapStyles()
        };
        if (this.googleMapsNode && !this.map) {
            this.map = new google.maps.Map(this.googleMapsNode, { ...mapOptions });
            // this.map.addListener("dragend", () => {
            //     this.setState({ dragged: true });
            // });
        } else if (this.map) {
            this.map.setOptions({ ...mapOptions });
        }
        this.setDefaultCenter(props);
    }

    private setDefaultCenter = (props: GoogleMapsProps) => {
        if (!props.fetchingData && props.allLocations && props.allLocations.length) {
            this.addMarkers(props.allLocations);
        }
    }

    private addMarkers = (mapLocations?: Location[]) => {
        this.markers.forEach(marker => marker.setMap(null));
        this.markers = [];
        if (mapLocations && mapLocations.length) {
            this.bounds = new google.maps.LatLngBounds();
            this.markers = mapLocations.reduce<google.maps.Marker[]>((markerArray, currentLocation) => {
                this.bounds.extend({
                    lat: currentLocation.latitude,
                    lng: currentLocation.longitude
                });
                const marker = new google.maps.Marker({
                    position: {
                        lat: currentLocation.latitude,
                        lng: currentLocation.longitude
                    },
                    icon: currentLocation.url
                });
                if (currentLocation.locationAttr && currentLocation.locationAttr.onClickEvent !== "doNothing") {
                    marker.addListener("click", (event: google.maps.MouseEvent) => {
                        if (this.props.onClickMarker && currentLocation.locationAttr) {
                            this.props.onClickMarker(event, currentLocation.locationAttr);
                        }
                    });
                } else {
                    marker.setClickable(false);
                }
                markerArray.push(marker);

                return markerArray;
            }, []);
            this.setMapOnMarkers(this.map);
            this.setBounds(this.bounds);
        }
    }

    private setBounds = (mapBounds?: google.maps.LatLngBounds) => {
        const { defaultCenterLatitude, defaultCenterLongitude, zoomLevel, autoZoom } = this.props;
        setTimeout(() => {
            if (mapBounds && this.map) {
                try {
                    this.map.fitBounds(mapBounds);
                    if (!autoZoom) {
                        if (defaultCenterLatitude && defaultCenterLongitude) {
                            this.map.setCenter({
                                lat: parseFloat(defaultCenterLatitude),
                                lng: parseFloat(defaultCenterLongitude)
                            });
                        }
                        this.map.setZoom(zoomLevel);
                    }
                } catch (error) {
                    this.setState({ alertMessage: `Invalid map bounds ${error.message}` });
                }
            }
        }, 0);
    }

    private setMapOnMarkers = (map?: google.maps.Map) => {
        if (this.markers && this.markers.length && map) {
            this.markers.forEach(marker => marker.setMap(map));
        }
    }

    private getMapStyles(): google.maps.MapTypeStyle[] {
        if (this.props.mapStyles && this.props.mapStyles.trim()) {
            try {
                return JSON.parse(this.props.mapStyles);
            } catch (error) {
                this.setState({ alertMessage: `invalid Map styles, ${error}` });
            }
        }

        return [ {
            featureType: "poi",
            elementType: "labels",
            stylers: [ { visibility: "off" } ]
        } ];
    }

}

export default googleApiWrapper(`https://maps.googleapis.com/maps/api/js?key=`)(GoogleMap);
