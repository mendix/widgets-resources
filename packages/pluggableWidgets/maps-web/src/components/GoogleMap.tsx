import { createElement, Dispatch, ReactElement, SetStateAction, useEffect, useRef, useState } from "react";
import classNames from "classnames";
import {
    GoogleMap as GoogleMapComponent,
    Marker as MarkerComponent,
    InfoWindow,
    useLoadScript
} from "@react-google-maps/api";
import { Marker, SharedProps } from "../../typings/shared";
import { getGoogleMapsStyles } from "../utils/google";
import { translateZoom } from "../utils/zoom";
import { Option } from "../utils/data";
import { Alert } from "@mendix/piw-utils-internal/components/web";
import { getDimensions } from "@mendix/piw-utils-internal";

export interface GoogleMapsProps extends SharedProps {
    mapStyles?: string;
    streetViewControl: boolean;
    mapTypeControl: boolean;
    fullscreenControl: boolean;
    rotateControl: boolean;
}

export function GoogleMap(props: GoogleMapsProps): ReactElement {
    const [map, setMap] = useState<google.maps.Map | undefined>();
    const center = useRef<google.maps.LatLngLiteral>({
        lat: 51.906688,
        lng: 4.48837
    });
    const [selectedMarker, setSelectedMarker] = useState<Option<Marker>>();
    const [error, setError] = useState("");
    const {
        autoZoom,
        className,
        currentLocation,
        fullscreenControl,
        locations,
        mapTypeControl,
        mapsToken,
        mapStyles,
        optionZoomControl: zoomControl,
        optionScroll: scrollwheel,
        optionDrag: draggable,
        rotateControl,
        streetViewControl,
        style,
        zoomLevel
    } = props;

    useEffect(() => {
        if (map) {
            const bounds = new google.maps.LatLngBounds();
            locations
                .concat(currentLocation ? [currentLocation] : [])
                .filter(m => !!m)
                .forEach(marker => {
                    bounds.extend({
                        lat: marker.latitude,
                        lng: marker.longitude
                    });
                });
            if (bounds.isEmpty()) {
                bounds.extend(center.current);
            }
            if (autoZoom) {
                map.fitBounds(bounds);
            } else {
                map.setCenter(bounds.getCenter());
            }
        }
    }, [map, locations, currentLocation, autoZoom]);

    const { isLoaded, loadError } = useLoadScript({
        googleMapsApiKey: mapsToken ?? "",
        id: "_com.mendix.widget.custom.Maps.Maps"
    });

    if (loadError) {
        setError(loadError.message);
    }

    return (
        <div className={classNames("widget-maps", className)} style={{ ...style, ...getDimensions(props) }}>
            {error && <Alert bootstrapStyle="danger">{error}</Alert>}
            <div className="widget-google-maps-wrapper">
                {isLoaded ? (
                    <GoogleMapComponent
                        mapContainerClassName="widget-google-maps"
                        options={{
                            draggable,
                            fullscreenControl,
                            mapTypeControl,
                            maxZoom: 20,
                            minZoom: 1,
                            rotateControl,
                            scrollwheel,
                            streetViewControl,
                            styles: getGoogleMapsStyles(mapStyles),
                            zoomControl
                        }}
                        onLoad={setMap}
                        onCenterChanged={() => {
                            if (map) {
                                center.current = map.getCenter().toJSON();
                            }
                        }}
                        zoom={autoZoom ? translateZoom("city") : zoomLevel}
                        center={center.current}
                    >
                        {locations
                            .concat(currentLocation ? [currentLocation] : [])
                            .filter(m => !!m)
                            .map((marker, index) => (
                                <GoogleMapsMarker
                                    key={`marker_${index}`}
                                    marker={marker}
                                    selectedMarker={selectedMarker}
                                    setSelectedMarker={setSelectedMarker}
                                />
                            ))}
                    </GoogleMapComponent>
                ) : (
                    <div className="spinner" />
                )}
            </div>
        </div>
    );
}

function GoogleMapsMarker({
    marker,
    selectedMarker,
    setSelectedMarker
}: {
    marker: Marker;
    selectedMarker: Option<Marker>;
    setSelectedMarker: Dispatch<SetStateAction<Option<Marker>>>;
}): ReactElement {
    const markerRef = useRef<google.maps.MVCObject>();
    return (
        <MarkerComponent
            position={{
                lat: marker.latitude,
                lng: marker.longitude
            }}
            title={marker.title}
            clickable={!!marker.title || !!marker.onClick}
            onLoad={ref => {
                markerRef.current = ref;
            }}
            onClick={
                marker.title ? () => setSelectedMarker(prev => (prev !== marker ? marker : undefined)) : marker.onClick
            }
            icon={marker.url}
        >
            {selectedMarker === marker && markerRef.current && (
                <InfoWindow
                    anchor={markerRef.current}
                    onCloseClick={() => setSelectedMarker(prev => (prev === marker ? undefined : prev))}
                >
                    <span style={{ cursor: marker.onClick ? "pointer" : "none" }} onClick={marker.onClick}>
                        {marker.title}
                    </span>
                </InfoWindow>
            )}
        </MarkerComponent>
    );
}
