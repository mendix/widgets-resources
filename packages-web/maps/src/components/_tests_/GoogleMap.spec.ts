import { mount, ReactWrapper, shallow, ShallowWrapper } from "enzyme";
import { createElement } from "react";
import { GoogleMap, GoogleMapsProps } from "../GoogleMap";
import { mockGoogleMaps } from "../../../tests/mocks/GoogleMaps";

describe("Google maps", () => {
    const defaultProps: GoogleMapsProps = {
        autoZoom: true,
        defaultCenterLatitude: "",
        defaultCenterLongitude: "",
        height: 75,
        heightUnit: "pixels",
        scriptsLoaded: true,
        optionDrag: true,
        optionScroll: true,
        optionZoomControl: true,
        zoomLevel: 10,
        width: 50,
        widthUnit: "percentage",
        mapStyles: "",
        divStyles: {},
        mapProvider: "googleMaps",
        inPreviewMode: false
    };

    beforeAll(() => {
        window.google = mockGoogleMaps;
    });

    const renderGoogleMap = (props: GoogleMapsProps): ShallowWrapper<GoogleMapsProps, any> =>
        shallow(createElement(GoogleMap, props));
    const fullRenderGoogleMap = (props: GoogleMapsProps): ReactWrapper<GoogleMapsProps, any> =>
        mount(createElement(GoogleMap, props));

    it("renders structure correctly", () => {
        const googleMaps = renderGoogleMap(defaultProps);
        googleMaps.setProps({
            heightUnit: "percentageOfWidth",
            widthUnit: "pixels"
        });

        expect(googleMaps).toMatchSnapshot();
    });

    it("with pixels renders structure correctly", () => {
        const googleMaps = renderGoogleMap(defaultProps);
        googleMaps.setProps({
            heightUnit: "pixels",
            widthUnit: "pixels"
        });

        expect(googleMaps).toMatchSnapshot();
    });

    it("with percentage of width and height units renders the structure correctly", () => {
        const googleMaps = renderGoogleMap(defaultProps);
        googleMaps.setProps({
            heightUnit: "percentageOfWidth",
            widthUnit: "percentage"
        });

        expect(googleMaps).toMatchSnapshot();
    });

    it("with percentage of parent units renders the structure correctly", () => {
        const googleMaps = renderGoogleMap(defaultProps);
        googleMaps.setProps({
            heightUnit: "percentageOfParent",
            widthUnit: "percentage"
        });

        expect(googleMaps).toMatchSnapshot();
    });

    it("without default center Latitude and Longitude sets default center location based on the default configured location", () => {
        const googleMaps = fullRenderGoogleMap(defaultProps);
        googleMaps.setProps({
            fetchingData: false
        });
        expect(googleMaps.state("center")).toEqual({ lat: 51.9066346, lng: 4.4861703 });
    });

    it("creates markers from given locations", () => {
        const customProps = {
            ...defaultProps,
            allLocations: [
                { latitude: 40.759011, longitude: -73.9844722, mxObject: undefined, url: "http://dummy.url" }
            ],
            fetchingData: false,
            autoZoom: false
        };
        const googleMaps = fullRenderGoogleMap(defaultProps);
        const googleMapsInstance = googleMaps.instance() as any;
        const createMarkerSpy = spyOn(googleMapsInstance, "addMarkers").and.callThrough();

        googleMapsInstance.componentWillReceiveProps(customProps);

        expect(createMarkerSpy).toHaveBeenCalledWith(customProps.allLocations);
    });

    afterAll(() => {
        window.google = undefined;
    });
});
