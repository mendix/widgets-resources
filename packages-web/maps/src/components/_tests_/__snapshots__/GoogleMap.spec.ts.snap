// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Google maps renders structure correctly 1`] = `
<div
  className="widget-maps"
  style={
    Object {
      "height": "37.5px",
      "width": "50px",
    }
  }
>
  <Alert
    bootstrapStyle="danger"
    className="widget-google-maps-alert"
  />
  <div
    className="widget-google-maps-wrapper"
  >
    <div
      className="widget-google-maps"
    />
  </div>
  <ResizeDetector
    handleHeight={true}
    handleWidth={true}
    nodeType="div"
    onResize={[Function]}
    querySelector={null}
    refreshMode="throttle"
    refreshRate={100}
    skipOnMount={false}
    targetDomEl={null}
  />
</div>
`;

exports[`Google maps with percentage of parent units renders the structure correctly 1`] = `
<div
  className="widget-maps"
  style={
    Object {
      "height": "75%",
      "width": "50%",
    }
  }
>
  <Alert
    bootstrapStyle="danger"
    className="widget-google-maps-alert"
  />
  <div
    className="widget-google-maps-wrapper"
  >
    <div
      className="widget-google-maps"
    />
  </div>
  <ResizeDetector
    handleHeight={true}
    handleWidth={true}
    nodeType="div"
    onResize={[Function]}
    querySelector={null}
    refreshMode="throttle"
    refreshRate={100}
    skipOnMount={false}
    targetDomEl={null}
  />
</div>
`;

exports[`Google maps with percentage of width and height units renders the structure correctly 1`] = `
<div
  className="widget-maps"
  style={
    Object {
      "height": "auto",
      "paddingBottom": "37.5%",
      "width": "50%",
    }
  }
>
  <Alert
    bootstrapStyle="danger"
    className="widget-google-maps-alert"
  />
  <div
    className="widget-google-maps-wrapper"
  >
    <div
      className="widget-google-maps"
    />
  </div>
  <ResizeDetector
    handleHeight={true}
    handleWidth={true}
    nodeType="div"
    onResize={[Function]}
    querySelector={null}
    refreshMode="throttle"
    refreshRate={100}
    skipOnMount={false}
    targetDomEl={null}
  />
</div>
`;

exports[`Google maps with pixels renders structure correctly 1`] = `
<div
  className="widget-maps"
  style={
    Object {
      "height": "75px",
      "width": "50px",
    }
  }
>
  <Alert
    bootstrapStyle="danger"
    className="widget-google-maps-alert"
  />
  <div
    className="widget-google-maps-wrapper"
  >
    <div
      className="widget-google-maps"
    />
  </div>
  <ResizeDetector
    handleHeight={true}
    handleWidth={true}
    nodeType="div"
    onResize={[Function]}
    querySelector={null}
    refreshMode="throttle"
    refreshRate={100}
    skipOnMount={false}
    targetDomEl={null}
  />
</div>
`;
